Automattic/woocommerce-payments
# Add/e2e add delete payment methods

Closes https://github.com/Automattic/woocommerce-payments/issues/2210 and https://github.com/Automattic/woocommerce-payments/issues/2209.

#### Changes proposed in this Pull Request

Title: New E2E Tests for Adding and Deleting Payment Methods in the 'My Account > Payment methods' page.

Description: This PR adds E2E tests that cover the following critical flows:
1. Shopper / Payment methods / Add new payment method
2. Shopper / Payment methods / Delete payment method  

#### Testing instructions

1. Setup and launch the e2e environment as usual.
2. Run these new e2e tests:
    ```
    npm run test:e2e ./tests/e2e/specs/shopper/shopper-myaccount-payment-methods-add-delete-success.spec.js
    npm run test:e2e ./tests/e2e/specs/shopper/shopper-myaccount-payment-methods-add-fail.spec.js
    ```
3. Verify that they pass like so:
  ```
   PASS  tests/e2e/specs/shopper/shopper-myaccount-payment-methods-add-delete-success.spec.js (111.604s)
   Payment Methods
      when using a basic card
          ✓ should add the card as a new payment method (17643ms)
          ✓ should be able to delete the card (5019ms)
      when using a 3ds card
          ✓ should add the card as a new payment method (21264ms)
          ✓ should be able to delete the card (5379ms)
      when using a 3ds2 card
          ✓ should add the card as a new payment method (19906ms)
          ✓ should be able to delete the card (5021ms)
  
  Test Suites: 1 passed, 1 total
  Tests:       6 passed, 6 total
  ```
  ```
 PASS  tests/e2e/specs/shopper/shopper-myaccount-payment-methods-add-fail.spec.js (101.883s)
  Payment Methods
    when attempting to add a declined card
      ✓ should not add the card (13408ms)
    when attempting to add a declined-funds card
      ✓ should not add the card (10971ms)
    when attempting to add a declined-incorrect card
      ✓ should not add the card (7100ms)
    when attempting to add a declined-expired card
      ✓ should not add the card (10916ms)
    when attempting to add a declined-cvc card
      ✓ should not add the card (10858ms)
    when attempting to add a declined-processing card
      ✓ should not add the card (10300ms)
    when attempting to add a declined-3ds card
      ✓ should not add the card (10350ms)

Test Suites: 1 passed, 1 total
Tests:       7 passed, 7 total
  ```

### Additional Notes

I discovered that there's a 20-second waiting time between adding payment methods. You would see how I implemented the waiting mechanism [here](https://github.com/Automattic/woocommerce-payments/pull/2392/files#diff-0b3eda82689e4d23b7692b27bc5afbcdbf7a0e96703c6b0069dffb8f2d7005e4R55). This limitation slows the test down so expect this test to be slower than the others.

-------------------
